File: explores/linked-list/0021_merge_two_sorted_lists.py
from typing import Optional

from ListNode import ListNode


class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dmy = ListNode(0)
        curr = dmy
        curr1 = list1
        curr2 = list2

        while curr1 and curr2:
            if curr1.val <= curr2.val:
                curr.next = curr1
                curr1 = curr1.next
            else:
                curr.next = curr2
                curr2 = curr2.next
            curr = curr.next

        if curr1:
            curr.next = curr1
        else:
            curr.next = curr2

        return dmy.next

File: explores/recursion-i/0021_merge_two_sorted_lists.cpp
struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
   public:
    ListNode* merge(ListNode* l1, ListNode* l2, ListNode* prev) {
        if (l1 == nullptr || l2 == nullptr) {
            return prev->next = l1 == nullptr ? l2 : l1;
        }

        if (l1->val <= l2->val) {
            prev->next = l1;
            l1 = l1->next;
        } else {
            prev->next = l2;
            l2 = l2->next;
        }
        prev = prev->next;
        return merge(l1, l2, prev);
    }
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* prev = new ListNode(-101);
        merge(list1, list2, prev);
        return prev->next;
    }
};